// @telekom/design-tokens/dist/telekom/tailwindcss-preset/telekom-design-tokens.config.js
/**
 * Do not edit directly
 * Generated on Thu, 04 May 2023 12:30:08 GMT
 *
 * You can find info about usage here:
 *   - https://github.com/telekom/design-tokens#how-to-use-the-tailwindcss-preset
 *   - https://tailwindcss.com/docs/presets
 */

const plugin = require('tailwindcss/plugin');

/**
 * Flatten shadow keys similarly to
 * `flattenColorPalette` utility function used by core color plugins
 */
const flattenShadows = (shadows) =>
  Object.assign(
    {},
    ...Object.entries(shadows ?? {}).flatMap(([shadow, values]) =>
      typeof values == 'object'
        ? Object.entries(flattenShadows(values)).map(([name, value]) => ({
            [shadow + (name === 'DEFAULT' ? '' : '-' + name)]: value,
          }))
        : [{ [shadow]: values }]
    )
  );

module.exports = {
  theme: {
    colors: {
      'text-&-icon': {
        DEFAULT: 'var(--telekom-color-text-and-icon-standard)',
        additional: 'var(--telekom-color-text-and-icon-additional)',
        disabled: 'var(--telekom-color-text-and-icon-disabled)',
        link: {
          DEFAULT: 'var(--telekom-color-text-and-icon-link-standard)',
          hovered: 'var(--telekom-color-text-and-icon-link-hovered)',
          visited: 'var(--telekom-color-text-and-icon-link-visited)',
          active: 'var(--telekom-color-text-and-icon-link-active)',
        },
        primary: {
          DEFAULT: 'var(--telekom-color-text-and-icon-primary-standard)',
          hovered: 'var(--telekom-color-text-and-icon-primary-hovered)',
          pressed: 'var(--telekom-color-text-and-icon-primary-pressed)',
        },
        inverted: {
          DEFAULT: 'var(--telekom-color-text-and-icon-inverted-standard)',
          additional: 'var(--telekom-color-text-and-icon-inverted-additional)',
        },
        white: {
          DEFAULT: 'var(--telekom-color-text-and-icon-white-standard)',
          additional: 'var(--telekom-color-text-and-icon-white-additional)',
        },
        black: {
          DEFAULT: 'var(--telekom-color-text-and-icon-black-standard)',
          additional: 'var(--telekom-color-text-and-icon-black-additional)',
        },
        functional: {
          informational:
            'var(--telekom-color-text-and-icon-functional-informational)',
          success: 'var(--telekom-color-text-and-icon-functional-success)',
          danger: 'var(--telekom-color-text-and-icon-functional-danger)',
          warning: 'var(--telekom-color-text-and-icon-functional-warning)',
        },
        'on-subtle': {
          informational:
            'var(--telekom-color-text-and-icon-on-subtle-informational)',
          success: 'var(--telekom-color-text-and-icon-on-subtle-success)',
          danger: 'var(--telekom-color-text-and-icon-on-subtle-danger)',
          warning: 'var(--telekom-color-text-and-icon-on-subtle-warning)',
          violet: 'var(--telekom-color-text-and-icon-on-subtle-violet)',
          blue: 'var(--telekom-color-text-and-icon-on-subtle-blue)',
          cyan: 'var(--telekom-color-text-and-icon-on-subtle-cyan)',
          teal: 'var(--telekom-color-text-and-icon-on-subtle-teal)',
          green: 'var(--telekom-color-text-and-icon-on-subtle-green)',
          olive: 'var(--telekom-color-text-and-icon-on-subtle-olive)',
          brown: 'var(--telekom-color-text-and-icon-on-subtle-brown)',
          yellow: 'var(--telekom-color-text-and-icon-on-subtle-yellow)',
          orange: 'var(--telekom-color-text-and-icon-on-subtle-orange)',
          red: 'var(--telekom-color-text-and-icon-on-subtle-red)',
        },
      },
      background: {
        canvas: 'var(--telekom-color-background-canvas)',
        'canvas-subtle': 'var(--telekom-color-background-canvas-subtle)',
        backdrop: 'var(--telekom-color-background-backdrop)',
        surface: 'var(--telekom-color-background-surface)',
        'surface-subtle': 'var(--telekom-color-background-surface-subtle)',
        'surface-highlight':
          'var(--telekom-color-background-surface-highlight)',
      },
      primary: {
        DEFAULT: 'var(--telekom-color-primary-standard)',
        hovered: 'var(--telekom-color-primary-hovered)',
        pressed: 'var(--telekom-color-primary-pressed)',
      },
      ui: {
        base: 'var(--telekom-color-ui-base)',
        subtle: 'var(--telekom-color-ui-subtle)',
        faint: 'var(--telekom-color-ui-faint)',
        regular: 'var(--telekom-color-ui-regular)',
        strong: 'var(--telekom-color-ui-strong)',
        'extra-strong': 'var(--telekom-color-ui-extra-strong)',
        white: 'var(--telekom-color-ui-white)',
        black: 'var(--telekom-color-ui-black)',
        disabled: 'var(--telekom-color-ui-disabled)',
        border: {
          DEFAULT: 'var(--telekom-color-ui-border-standard)',
          hovered: 'var(--telekom-color-ui-border-hovered)',
          pressed: 'var(--telekom-color-ui-border-pressed)',
          disabled: 'var(--telekom-color-ui-border-disabled)',
        },
        'state-fill': {
          DEFAULT: 'var(--telekom-color-ui-state-fill-standard)',
          hovered: 'var(--telekom-color-ui-state-fill-hovered)',
          pressed: 'var(--telekom-color-ui-state-fill-pressed)',
          'hovered-inverted':
            'var(--telekom-color-ui-state-fill-hovered-inverted)',
          'pressed-inverted':
            'var(--telekom-color-ui-state-fill-pressed-inverted)',
        },
      },
      functional: {
        focus: {
          DEFAULT: 'var(--telekom-color-functional-focus-standard)',
          inverted: 'var(--telekom-color-functional-focus-inverted)',
          'on-dark-background':
            'var(--telekom-color-functional-focus-on-dark-background)',
          'on-light-background':
            'var(--telekom-color-functional-focus-on-light-background)',
        },
        informational: {
          DEFAULT: 'var(--telekom-color-functional-informational-standard)',
          hovered: 'var(--telekom-color-functional-informational-hovered)',
          pressed: 'var(--telekom-color-functional-informational-pressed)',
          subtle: 'var(--telekom-color-functional-informational-subtle)',
        },
        success: {
          DEFAULT: 'var(--telekom-color-functional-success-standard)',
          hovered: 'var(--telekom-color-functional-success-hovered)',
          pressed: 'var(--telekom-color-functional-success-pressed)',
          subtle: 'var(--telekom-color-functional-success-subtle)',
        },
        danger: {
          DEFAULT: 'var(--telekom-color-functional-danger-standard)',
          hovered: 'var(--telekom-color-functional-danger-hovered)',
          pressed: 'var(--telekom-color-functional-danger-pressed)',
          subtle: 'var(--telekom-color-functional-danger-subtle)',
        },
        warning: {
          DEFAULT: 'var(--telekom-color-functional-warning-standard)',
          hovered: 'var(--telekom-color-functional-warning-hovered)',
          pressed: 'var(--telekom-color-functional-warning-pressed)',
          subtle: 'var(--telekom-color-functional-warning-subtle)',
        },
      },
      additional: {
        violet: {
          0: 'var(--telekom-color-additional-violet-0)',
          100: 'var(--telekom-color-additional-violet-100)',
          200: 'var(--telekom-color-additional-violet-200)',
          300: 'var(--telekom-color-additional-violet-300)',
          400: 'var(--telekom-color-additional-violet-400)',
          500: 'var(--telekom-color-additional-violet-500)',
          600: 'var(--telekom-color-additional-violet-600)',
          700: 'var(--telekom-color-additional-violet-700)',
          800: 'var(--telekom-color-additional-violet-800)',
          900: 'var(--telekom-color-additional-violet-900)',
          subtle: 'var(--telekom-color-additional-violet-subtle)',
        },
        blue: {
          0: 'var(--telekom-color-additional-blue-0)',
          100: 'var(--telekom-color-additional-blue-100)',
          200: 'var(--telekom-color-additional-blue-200)',
          300: 'var(--telekom-color-additional-blue-300)',
          400: 'var(--telekom-color-additional-blue-400)',
          500: 'var(--telekom-color-additional-blue-500)',
          600: 'var(--telekom-color-additional-blue-600)',
          700: 'var(--telekom-color-additional-blue-700)',
          800: 'var(--telekom-color-additional-blue-800)',
          900: 'var(--telekom-color-additional-blue-900)',
          subtle: 'var(--telekom-color-additional-blue-subtle)',
        },
        cyan: {
          0: 'var(--telekom-color-additional-cyan-0)',
          100: 'var(--telekom-color-additional-cyan-100)',
          200: 'var(--telekom-color-additional-cyan-200)',
          300: 'var(--telekom-color-additional-cyan-300)',
          400: 'var(--telekom-color-additional-cyan-400)',
          500: 'var(--telekom-color-additional-cyan-500)',
          600: 'var(--telekom-color-additional-cyan-600)',
          700: 'var(--telekom-color-additional-cyan-700)',
          800: 'var(--telekom-color-additional-cyan-800)',
          900: 'var(--telekom-color-additional-cyan-900)',
          subtle: 'var(--telekom-color-additional-cyan-subtle)',
        },
        teal: {
          0: 'var(--telekom-color-additional-teal-0)',
          100: 'var(--telekom-color-additional-teal-100)',
          200: 'var(--telekom-color-additional-teal-200)',
          300: 'var(--telekom-color-additional-teal-300)',
          400: 'var(--telekom-color-additional-teal-400)',
          500: 'var(--telekom-color-additional-teal-500)',
          600: 'var(--telekom-color-additional-teal-600)',
          700: 'var(--telekom-color-additional-teal-700)',
          800: 'var(--telekom-color-additional-teal-800)',
          900: 'var(--telekom-color-additional-teal-900)',
          subtle: 'var(--telekom-color-additional-teal-subtle)',
        },
        green: {
          0: 'var(--telekom-color-additional-green-0)',
          100: 'var(--telekom-color-additional-green-100)',
          200: 'var(--telekom-color-additional-green-200)',
          300: 'var(--telekom-color-additional-green-300)',
          400: 'var(--telekom-color-additional-green-400)',
          500: 'var(--telekom-color-additional-green-500)',
          600: 'var(--telekom-color-additional-green-600)',
          700: 'var(--telekom-color-additional-green-700)',
          800: 'var(--telekom-color-additional-green-800)',
          900: 'var(--telekom-color-additional-green-900)',
          subtle: 'var(--telekom-color-additional-green-subtle)',
        },
        olive: {
          0: 'var(--telekom-color-additional-olive-0)',
          100: 'var(--telekom-color-additional-olive-100)',
          200: 'var(--telekom-color-additional-olive-200)',
          300: 'var(--telekom-color-additional-olive-300)',
          400: 'var(--telekom-color-additional-olive-400)',
          500: 'var(--telekom-color-additional-olive-500)',
          600: 'var(--telekom-color-additional-olive-600)',
          700: 'var(--telekom-color-additional-olive-700)',
          800: 'var(--telekom-color-additional-olive-800)',
          900: 'var(--telekom-color-additional-olive-900)',
          subtle: 'var(--telekom-color-additional-olive-subtle)',
        },
        brown: {
          0: 'var(--telekom-color-additional-brown-0)',
          100: 'var(--telekom-color-additional-brown-100)',
          200: 'var(--telekom-color-additional-brown-200)',
          300: 'var(--telekom-color-additional-brown-300)',
          400: 'var(--telekom-color-additional-brown-400)',
          500: 'var(--telekom-color-additional-brown-500)',
          600: 'var(--telekom-color-additional-brown-600)',
          700: 'var(--telekom-color-additional-brown-700)',
          800: 'var(--telekom-color-additional-brown-800)',
          900: 'var(--telekom-color-additional-brown-900)',
          subtle: 'var(--telekom-color-additional-brown-subtle)',
        },
        yellow: {
          0: 'var(--telekom-color-additional-yellow-0)',
          100: 'var(--telekom-color-additional-yellow-100)',
          200: 'var(--telekom-color-additional-yellow-200)',
          300: 'var(--telekom-color-additional-yellow-300)',
          400: 'var(--telekom-color-additional-yellow-400)',
          500: 'var(--telekom-color-additional-yellow-500)',
          600: 'var(--telekom-color-additional-yellow-600)',
          700: 'var(--telekom-color-additional-yellow-700)',
          800: 'var(--telekom-color-additional-yellow-800)',
          900: 'var(--telekom-color-additional-yellow-900)',
          subtle: 'var(--telekom-color-additional-yellow-subtle)',
        },
        orange: {
          0: 'var(--telekom-color-additional-orange-0)',
          100: 'var(--telekom-color-additional-orange-100)',
          200: 'var(--telekom-color-additional-orange-200)',
          300: 'var(--telekom-color-additional-orange-300)',
          400: 'var(--telekom-color-additional-orange-400)',
          500: 'var(--telekom-color-additional-orange-500)',
          600: 'var(--telekom-color-additional-orange-600)',
          700: 'var(--telekom-color-additional-orange-700)',
          800: 'var(--telekom-color-additional-orange-800)',
          900: 'var(--telekom-color-additional-orange-900)',
          subtle: 'var(--telekom-color-additional-orange-subtle)',
        },
        red: {
          0: 'var(--telekom-color-additional-red-0)',
          100: 'var(--telekom-color-additional-red-100)',
          200: 'var(--telekom-color-additional-red-200)',
          300: 'var(--telekom-color-additional-red-300)',
          400: 'var(--telekom-color-additional-red-400)',
          500: 'var(--telekom-color-additional-red-500)',
          600: 'var(--telekom-color-additional-red-600)',
          700: 'var(--telekom-color-additional-red-700)',
          800: 'var(--telekom-color-additional-red-800)',
          900: 'var(--telekom-color-additional-red-900)',
          subtle: 'var(--telekom-color-additional-red-subtle)',
        },
      },
    },
    borderWidth: {
      DEFAULT: 'var(--telekom-line-weight-standard)',
      highlight: 'var(--telekom-line-weight-highlight)',
      bold: 'var(--telekom-line-weight-bold)',
    },
    transitionDuration: {
      instant: 'var(--telekom-motion-duration-instant)',
      immediate: 'var(--telekom-motion-duration-immediate)',
      transition: 'var(--telekom-motion-duration-transition)',
      animation: 'var(--telekom-motion-duration-animation)',
      'animation-deliberate':
        'var(--telekom-motion-duration-animation-deliberate)',
    },
    transitionTimingFunction: {
      DEFAULT: 'var(--telekom-motion-easing-standard)',
      enter: 'var(--telekom-motion-easing-enter)',
      leave: 'var(--telekom-motion-easing-leave)',
    },
    opacity: {
      invisible: 'var(--telekom-opacity-invisible)',
      transparent: 'var(--telekom-opacity-transparent)',
      'semi-transparent': 'var(--telekom-opacity-semi-transparent)',
      translucent: 'var(--telekom-opacity-translucent)',
      'semi-translucent': 'var(--telekom-opacity-semi-translucent)',
      solid: 'var(--telekom-opacity-solid)',
    },
    borderRadius: {
      small: 'var(--telekom-radius-small)',
      DEFAULT: 'var(--telekom-radius-standard)',
      large: 'var(--telekom-radius-large)',
      pill: 'var(--telekom-radius-pill)',
      circle: 'var(--telekom-radius-circle)',
    },
    shadow: {
      raised: {
        DEFAULT: 'var(--telekom-shadow-raised-standard)',
        hover: 'var(--telekom-shadow-raised-hover)',
        pressed: 'var(--telekom-shadow-raised-pressed)',
      },
      floating: {
        DEFAULT: 'var(--telekom-shadow-floating-standard)',
        hover: 'var(--telekom-shadow-floating-hover)',
        pressed: 'var(--telekom-shadow-floating-pressed)',
      },
      top: 'var(--telekom-shadow-top)',
      overlay: 'var(--telekom-shadow-overlay)',
      'app-bar': {
        top: {
          raised: 'var(--telekom-shadow-app-bar-top-raised)',
          flat: 'var(--telekom-shadow-app-bar-top-flat)',
        },
        bottom: {
          raised: 'var(--telekom-shadow-app-bar-bottom-raised)',
          flat: 'var(--telekom-shadow-app-bar-bottom-flat)',
        },
      },
    },
    spacing: {
      0: 'var(--telekom-spacing-unit-x0)',
      1: 'var(--telekom-spacing-unit-x1)',
      2: 'var(--telekom-spacing-unit-x2)',
      3: 'var(--telekom-spacing-unit-x3)',
      4: 'var(--telekom-spacing-unit-x4)',
      5: 'var(--telekom-spacing-unit-x5)',
      6: 'var(--telekom-spacing-unit-x6)',
      7: 'var(--telekom-spacing-unit-x7)',
      8: 'var(--telekom-spacing-unit-x8)',
      9: 'var(--telekom-spacing-unit-x9)',
      10: 'var(--telekom-spacing-unit-x10)',
      11: 'var(--telekom-spacing-unit-x11)',
      12: 'var(--telekom-spacing-unit-x12)',
      13: 'var(--telekom-spacing-unit-x13)',
      14: 'var(--telekom-spacing-unit-x14)',
      15: 'var(--telekom-spacing-unit-x15)',
      16: 'var(--telekom-spacing-unit-x16)',
      17: 'var(--telekom-spacing-unit-x17)',
      18: 'var(--telekom-spacing-unit-x18)',
      19: 'var(--telekom-spacing-unit-x19)',
      20: 'var(--telekom-spacing-unit-x20)',
      '025': 'var(--telekom-spacing-unit-x025)',
      '05': 'var(--telekom-spacing-unit-x05)',
    },
    textStyle: {
      badge: 'var(--telekom-text-style-badge)',
      small: 'var(--telekom-text-style-small)',
      'small-bold': 'var(--telekom-text-style-small-bold)',
      caption: 'var(--telekom-text-style-caption)',
      'caption-bold': 'var(--telekom-text-style-caption-bold)',
      body: 'var(--telekom-text-style-body)',
      'body-bold': 'var(--telekom-text-style-body-bold)',
      ui: 'var(--telekom-text-style-ui)',
      'ui-bold': 'var(--telekom-text-style-ui-bold)',
      'lead-text': 'var(--telekom-text-style-lead-text)',
      'heading-6': 'var(--telekom-text-style-heading-6)',
      'heading-5': 'var(--telekom-text-style-heading-5)',
      'heading-4': 'var(--telekom-text-style-heading-4)',
      'heading-3': 'var(--telekom-text-style-heading-3)',
      'heading-2': 'var(--telekom-text-style-heading-2)',
      'heading-1': 'var(--telekom-text-style-heading-1)',
      'title-2': 'var(--telekom-text-style-title-2)',
      'title-1': 'var(--telekom-text-style-title-1)',
    },
    fontSize: {
      badge: 'var(--telekom-typography-font-size-badge)',
      small: 'var(--telekom-typography-font-size-small)',
      caption: 'var(--telekom-typography-font-size-caption)',
      body: 'var(--telekom-typography-font-size-body)',
      callout: 'var(--telekom-typography-font-size-callout)',
      'headline-3': 'var(--telekom-typography-font-size-headline-3)',
      'headline-2': 'var(--telekom-typography-font-size-headline-2)',
      'headline-1': 'var(--telekom-typography-font-size-headline-1)',
      'title-3': 'var(--telekom-typography-font-size-title-3)',
      'title-2': 'var(--telekom-typography-font-size-title-2)',
      'title-1': 'var(--telekom-typography-font-size-title-1)',
    },
    fontFamily: {
      sans: 'var(--telekom-typography-font-family-sans)',
      mono: 'var(--telekom-typography-font-family-mono)',
    },
    fontWeight: {
      thin: 'var(--telekom-typography-font-weight-thin)',
      regular: 'var(--telekom-typography-font-weight-regular)',
      medium: 'var(--telekom-typography-font-weight-medium)',
      bold: 'var(--telekom-typography-font-weight-bold)',
      'extra-bold': 'var(--telekom-typography-font-weight-extra-bold)',
    },
    lineHeight: {
      none: 'var(--telekom-typography-line-spacing-none)',
      'extra-tight': 'var(--telekom-typography-line-spacing-extra-tight)',
      tight: 'var(--telekom-typography-line-spacing-tight)',
      DEFAULT: 'var(--telekom-typography-line-spacing-standard)',
      loose: 'var(--telekom-typography-line-spacing-loose)',
    },
    letterSpacing: {
      DEFAULT: 'var(--telekom-typography-letter-spacing-standard)',
    },
  },
  plugins: [
    /**
     * Custom plugin to convert `text-style` design token to the `font` css shorthand property
     * as there is no mapping provided by the core plugins.
     */
    plugin(function ({ matchUtilities, theme }) {
      matchUtilities(
        {
          'text-style': (value) => ({
            font: value,
          }),
        },
        { values: theme('textStyle') }
      );
    }),
    /**
     * Custom plugin to convert `shadow` design token to the `boxShadow` css shorthand property
     * and to avoid functionality of core Tailwind CSS plugins (`boxShadow` and `boxShadowColor`).
     */
    plugin(function ({ matchUtilities, theme }) {
      matchUtilities(
        {
          shadow: (value) => ({
            boxShadow: value,
          }),
        },
        { values: flattenShadows(theme('shadow')) }
      );
    }),
  ],
  corePlugins: {
    boxShadow: false,
    boxShadowColor: false,
  },
};
